File: web/shared/pricing/catalog.ts
import catalogData from "./tiers.json";

export type WorkspacePlan = "personal" | "pro_trial" | "pro" | "enterprise";
export type PlanFamily = "free" | "pro" | "enterprise";
export type BillingState = "none" | "trial" | "paid" | "sales";
export type Availability = PlanFamily | "separate_add_on" | "not_yet_publicly_tiered";

export type WorkspacePlanMapping = {
  planFamily: PlanFamily;
  billingState: BillingState;
};

export type PlanFamilyDefinition = {
  label: string;
  comparisonLabel: string;
  price: string;
  cadence: string;
  description: string;
  ctaLabel: string;
  waitlistLabel: string;
  loginPlan?: WorkspacePlan;
  contactHref?: string;
  featured?: boolean;
  bullets: string[];
  entitlements: string[];
};

export type PricingMatrixRow = {
  key: string;
  label: string;
  free: string;
  pro: string;
  enterprise: string;
};

export type PricingMatrixGroup = {
  title: string;
  rows: PricingMatrixRow[];
};

export type FAQ = {
  question: string;
  answer: string;
};

export type FeatureInventoryItem = {
  key: string;
  label: string;
  category: string;
  availability: Availability;
  status: "implemented" | "service" | "internal";
};

export type PricingCatalog = {
  version: number;
  workspacePlanMappings: Record<WorkspacePlan, WorkspacePlanMapping>;
  publicPlanOrder: PlanFamily[];
  planFamilies: Record<PlanFamily, PlanFamilyDefinition>;
  pricingMatrix: PricingMatrixGroup[];
  faqs: FAQ[];
  featureInventory: FeatureInventoryItem[];
  separateAddOns: Array<{ key: string; label: string }>;
  notYetPubliclyTiered: Array<{ key: string; label: string }>;
};

export type WorkspacePlanProfile = {
  workspacePlan: WorkspacePlan;
  planFamily: PlanFamily;
  billingState: BillingState;
  label: string;
  entitlements: string[];
};

export const pricingCatalog = catalogData as PricingCatalog;

export function normalizeWorkspacePlan(raw: string | null | undefined): WorkspacePlan {
  const value = (raw ?? "").trim().toLowerCase();
  if (value === "free" || value === "personal") return "personal";
  if (value === "pro_trial") return "pro_trial";
  if (value === "pro") return "pro";
  if (value === "enterprise") return "enterprise";
  return "personal";
}

export function workspacePlanProfile(raw: string | null | undefined): WorkspacePlanProfile {
  const workspacePlan = normalizeWorkspacePlan(raw);
  const mapping = pricingCatalog.workspacePlanMappings[workspacePlan];
  const family = pricingCatalog.planFamilies[mapping.planFamily];
  return {
    workspacePlan,
    planFamily: mapping.planFamily,
    billingState: mapping.billingState,
    label: family.label,
    entitlements: [...family.entitlements],
  };
}

export function workspacePlanLabel(raw: string | null | undefined): string {
  return workspacePlanProfile(raw).label;
}

export function hasEntitlement(entitlements: readonly string[] | undefined, key: string): boolean {
  if (!entitlements || !key.trim()) return false;
  return entitlements.some((value) => value.trim() === key.trim());
}

export function publicPlanDefinitions(): Array<PlanFamilyDefinition & { id: PlanFamily }> {
  return pricingCatalog.publicPlanOrder.map((id) => ({
    id,
    ...pricingCatalog.planFamilies[id],
  }));
}

File: web/shared/pricing/catalog.go
package pricing

import (
	_ "embed"
	"encoding/json"
	"strings"
)

const (
	EntitlementQuickAdd              = "quick_add"
	EntitlementRecurringSchedules    = "recurring_schedules"
	EntitlementPersonalBoardGameplay = "personal_board_gameplay"
	EntitlementCalendarSync          = "calendar_sync"
	EntitlementPWAOffline            = "pwa_offline"
	EntitlementSharedTeamBoards      = "shared_team_boards"
	EntitlementWorkspaceInvites      = "workspace_invites"
	EntitlementTeamRoles             = "team_roles"
	EntitlementTeamAdmin             = "team_admin"
	EntitlementBoardMemberManagement = "board_member_management"
	EntitlementGuidedMigration       = "guided_migration"
	EntitlementSecurityReviewSupport = "security_review_support"
	EntitlementProcurementSupport    = "procurement_support"
	EntitlementPriorityMigration     = "priority_migration_support"
)

type WorkspacePlanMapping struct {
	PlanFamily   string `json:"planFamily"`
	BillingState string `json:"billingState"`
}

type PlanFamily struct {
	Label           string   `json:"label"`
	ComparisonLabel string   `json:"comparisonLabel"`
	Price           string   `json:"price"`
	Cadence         string   `json:"cadence"`
	Description     string   `json:"description"`
	CTALabel        string   `json:"ctaLabel"`
	WaitlistLabel   string   `json:"waitlistLabel"`
	LoginPlan       string   `json:"loginPlan"`
	ContactHref     string   `json:"contactHref"`
	Featured        bool     `json:"featured"`
	Bullets         []string `json:"bullets"`
	Entitlements    []string `json:"entitlements"`
}

type PricingMatrixRow struct {
	Key        string `json:"key"`
	Label      string `json:"label"`
	Free       string `json:"free"`
	Pro        string `json:"pro"`
	Enterprise string `json:"enterprise"`
}

type PricingMatrixGroup struct {
	Title string             `json:"title"`
	Rows  []PricingMatrixRow `json:"rows"`
}

type FAQ struct {
	Question string `json:"question"`
	Answer   string `json:"answer"`
}

type FeatureInventoryItem struct {
	Key          string `json:"key"`
	Label        string `json:"label"`
	Category     string `json:"category"`
	Availability string `json:"availability"`
	Status       string `json:"status"`
}

type ListItem struct {
	Key   string `json:"key"`
	Label string `json:"label"`
}

type Catalog struct {
	Version               int                             `json:"version"`
	WorkspacePlanMappings map[string]WorkspacePlanMapping `json:"workspacePlanMappings"`
	PublicPlanOrder       []string                        `json:"publicPlanOrder"`
	PlanFamilies          map[string]PlanFamily           `json:"planFamilies"`
	PricingMatrix         []PricingMatrixGroup            `json:"pricingMatrix"`
	FAQs                  []FAQ                           `json:"faqs"`
	FeatureInventory      []FeatureInventoryItem          `json:"featureInventory"`
	SeparateAddOns        []ListItem                      `json:"separateAddOns"`
	NotYetPubliclyTiered  []ListItem                      `json:"notYetPubliclyTiered"`
}

type WorkspacePlanProfile struct {
	WorkspacePlan string
	PlanFamily    string
	BillingState  string
	DisplayLabel  string
	Entitlements  []string
}

//go:embed tiers.json
var rawCatalog []byte

var catalog = mustLoadCatalog()

func CatalogData() Catalog {
	return cloneCatalog(catalog)
}

func LookupWorkspacePlan(raw string) WorkspacePlanProfile {
	normalized := NormalizeWorkspacePlan(raw)
	mapping, ok := catalog.WorkspacePlanMappings[normalized]
	if !ok {
		normalized = "personal"
		mapping = catalog.WorkspacePlanMappings[normalized]
	}
	family := catalog.PlanFamilies[mapping.PlanFamily]
	return WorkspacePlanProfile{
		WorkspacePlan: normalized,
		PlanFamily:    mapping.PlanFamily,
		BillingState:  mapping.BillingState,
		DisplayLabel:  family.Label,
		Entitlements:  append([]string{}, family.Entitlements...),
	}
}

func NormalizeWorkspacePlan(raw string) string {
	switch strings.ToLower(strings.TrimSpace(raw)) {
	case "free", "personal":
		return "personal"
	case "pro_trial":
		return "pro_trial"
	case "pro":
		return "pro"
	case "enterprise":
		return "enterprise"
	default:
		return "personal"
	}
}

func HasEntitlement(entitlements []string, key string) bool {
	needle := strings.TrimSpace(key)
	if needle == "" {
		return false
	}
	for _, entitlement := range entitlements {
		if strings.TrimSpace(entitlement) == needle {
			return true
		}
	}
	return false
}

func mustLoadCatalog() Catalog {
	var parsed Catalog
	if err := json.Unmarshal(rawCatalog, &parsed); err != nil {
		panic(err)
	}
	return cloneCatalog(parsed)
}

func cloneCatalog(source Catalog) Catalog {
	cloned := source
	cloned.PublicPlanOrder = append([]string{}, source.PublicPlanOrder...)
	cloned.PlanFamilies = make(map[string]PlanFamily, len(source.PlanFamilies))
	for key, family := range source.PlanFamilies {
		next := family
		next.Bullets = append([]string{}, family.Bullets...)
		next.Entitlements = append([]string{}, family.Entitlements...)
		cloned.PlanFamilies[key] = next
	}
	cloned.WorkspacePlanMappings = make(map[string]WorkspacePlanMapping, len(source.WorkspacePlanMappings))
	for key, mapping := range source.WorkspacePlanMappings {
		cloned.WorkspacePlanMappings[key] = mapping
	}
	cloned.PricingMatrix = make([]PricingMatrixGroup, 0, len(source.PricingMatrix))
	for _, group := range source.PricingMatrix {
		next := group
		next.Rows = append([]PricingMatrixRow{}, group.Rows...)
		cloned.PricingMatrix = append(cloned.PricingMatrix, next)
	}
	cloned.FAQs = append([]FAQ{}, source.FAQs...)
	cloned.FeatureInventory = append([]FeatureInventoryItem{}, source.FeatureInventory...)
	cloned.SeparateAddOns = append([]ListItem{}, source.SeparateAddOns...)
	cloned.NotYetPubliclyTiered = append([]ListItem{}, source.NotYetPubliclyTiered...)
	return cloned
}
